### PR TITLE
Port :filename.basename/1 to JS

### PR DESCRIPTION
**Pull request by danirukun**
Closes https://github.com/bartblast/hologram/issues/324

### CODE REVIEW

**Comment by bartblast** [comment]
I've just added "Quality Checks Before Submitting" section here: https://hologram.page/reference/contributing, though to get the updated checks you'll need to sync with the `dev` branch.

File: assets/js/erlang/filename.mjs [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+  "basename/1": (atomOrBinaryOrList) => {
+    if (Type.isList(atomOrBinaryOrList)) {
+      if (atomOrBinaryOrList.data.length === 0) {
+        return Type.list([]);
+      }
+
+      const binary = Erlang["iolist_to_binary/1"](atomOrBinaryOrList);
+      Bitstring.maybeSetTextFromBytes(binary);
+      const parts = binary.text.split("/").filter((part) => part !== "");
+      const basenameText = parts.length > 0 ? parts[parts.length - 1] : "";
```
Comment (bartblast) [inline assets/js/erlang/filename.mjs]: `parts[parts.length - 1]`
->
`parts.at(-1)`

File: assets/js/erlang/filename.mjs [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+      const basenameBinary = Type.bitstring(basenameText);
+      return Bitstring.toCodepoints(basenameBinary);
+    }
+
+    if (
```
Comment (bartblast) [inline assets/js/erlang/filename.mjs]: !Type.isBinary() alone is enough

File: assets/js/erlang/filename.mjs [hunk 3]
```diff
@@ -0,0 +1,67 @@
+"use strict";
+
+import Bitstring from "../bitstring.mjs";
+import Interpreter from "../interpreter.mjs";
+import Type from "../type.mjs";
+
+const Erlang_Filename = {
+  // Start basename/1
+  "basename/1": (atomOrBinaryOrList) => {
```
Comment (bartblast) [inline assets/js/erlang/filename.mjs]: IMO you can substantially simplify the code if you write it this way:

```js
let filepathText;

if (Type.isBinary(filepath)) { // this will be the most common case
   // binary-specific behaviour and conversion
  ...
  filepathText = ...;
} else if (Type.isList(term) {
  ...
} else if (Type.isAtom(term) {
  ...
} else {
  // here we're raising the exception
}

const parts = filepathText.split("/").filter((part) => part !== "");
...
```

Comment (bartblast) [inline assets/js/erlang/filename.mjs]: Could you use `filename` instead of `atomOrBinaryOrList`?

File: assets/js/erlang/filename.mjs [hunk 4]
```diff
@@ -0,0 +1,67 @@
+"use strict";
+
+import Bitstring from "../bitstring.mjs";
+import Interpreter from "../interpreter.mjs";
+import Type from "../type.mjs";
+
```
Comment (bartblast) [inline assets/js/erlang/filename.mjs:11]: Please keep the preamble:

```
// IMPORTANT!
// If the given ported Erlang function calls other Erlang functions, then list such dependencies in the "Deps" comment (see :erlang./=/2 for an example).
// Also, in such case add respective call graph edges in Hologram.CallGraph.list_runtime_mfas/1.
```

File: test/elixir/hologram/ex_js_consistency/erlang/filename_test.exs [hunk 5]
```diff
@@ -0,0 +1,109 @@
+defmodule Hologram.ExJsConsistency.Erlang.FilenameTest do
+  @moduledoc """
+  IMPORTANT!
+  Each Elixir consistency test has a related JavaScript test in test/javascript/erlang/filename_test.mjs
+  Always update both together.
+  """
```
Comment (bartblast) [inline test/elixir/hologram/ex_js_consistency/erlang/filename_test.exs:6]: Add newline between, please.

File: test/elixir/hologram/ex_js_consistency/erlang/filename_test.exs [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+  use Hologram.Test.BasicCase, async: true
+
+  @moduletag :consistency
+
+  describe "basename/1" do
+    test "returns basename for path with multiple segments" do
```
Comment (bartblast) [inline test/elixir/hologram/ex_js_consistency/erlang/filename_test.exs]: "returns basename for " prefix for all of these tests is redundant IMO,
"path with multiple segments" is enough.

File: test/elixir/hologram/ex_js_consistency/erlang/filename_test.exs [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+
+    test "returns basename for path with unicode characters" do
+      assert :filename.basename("path/to/文件.txt") == "文件.txt"
+    end
+
+    test "returns basename for path with special characters" do
+      assert :filename.basename("path/to/file-name_123.txt") == "file-name_123.txt"
+    end
+
+    test "returns basename as list of code points for atom input" do
```
Comment (bartblast) [inline test/elixir/hologram/ex_js_consistency/erlang/filename_test.exs]: One test for non-nil value and one for nil is enough I think...

File: test/elixir/hologram/ex_js_consistency/erlang/filename_test.exs [hunk 8]
```diff
@@ -0,0 +1,109 @@
...
+
+    test "returns basename as list of code points for nil atom input with path" do
+      assert :filename.basename(:"path/to/nil") == ~c"nil"
+    end
+
+    test "returns basename as list of code points for atom input with unicode characters" do
+      assert :filename.basename(:"path/to/文件.txt") == ~c"文件.txt"
+    end
+
+    test "returns empty list for empty list input" do
```
Comment (bartblast) [inline test/elixir/hologram/ex_js_consistency/erlang/filename_test.exs:57]: Similar to atoms, testing one empty list and one non-empty list is sufficient. Otherwise, we're essentially testing `:erlang.atom_to_binary/1` and `:erlang.iolist_to_binary/1` themselves.
(This applies if you simplify the function as I outlined initially.)

File: test/elixir/hologram/ex_js_consistency/erlang/filename_test.exs [hunk 9]
```diff
@@ -0,0 +1,109 @@
...
+
+    test "returns basename as list of code points for non-empty list input" do
+      assert :filename.basename([~c"path/to/", ?f, ?i, ?l, ?e, ~c".txt"]) == ~c"file.txt"
+    end
+
+    test "returns basename as list of code points for list input with single segment" do
+      assert :filename.basename([?f, ?i, ?l, ?e, ?n, ?a, ?m, ?e]) == ~c"filename"
+    end
+
+    test "raises FunctionClauseError if the argument is not a binary or atom" do
```
Comment (bartblast) [inline test/elixir/hologram/ex_js_consistency/erlang/filename_test.exs]: ... or list

File: test/elixir/hologram/ex_js_consistency/erlang/filename_test.exs [hunk 10]
```diff
@@ -0,0 +1,109 @@
...
+      assert_error FunctionClauseError,
+                   build_function_clause_error_msg(":filename.do_flatten/2", [123, []]),
+                   fn -> :filename.basename(123) end
+    end
+
+    test "raises FunctionClauseError if the argument is a tuple" do
```
Comment (bartblast) [inline test/elixir/hologram/ex_js_consistency/erlang/filename_test.exs]: This is redundant, otherwise we'd need to test all other types.

File: test/elixir/hologram/ex_js_consistency/erlang/filename_test.exs [hunk 11]
```diff
@@ -0,0 +1,109 @@
...
+      assert_error FunctionClauseError,
+                   build_function_clause_error_msg(":filename.do_flatten/2", [{1, 2}, []]),
+                   fn -> :filename.basename({1, 2}) end
+    end
+
+    test "raises FunctionClauseError if the argument is a non-binary bitstring" do
```
Comment (bartblast) [inline test/elixir/hologram/ex_js_consistency/erlang/filename_test.exs:75]: However, this one is worth keeping.

File: test/javascript/erlang/filename_test.mjs [hunk 12]
```diff
@@ -0,0 +1,225 @@
...
+      const result = basename(Type.bitstring("path/to/file-name_123.txt"));
+      const expected = Type.bitstring("file-name_123.txt");
+
+      assert.deepStrictEqual(result, expected);
+    });
+
+    it("returns basename as list of code points for atom input", () => {
+      const result = basename(Type.atom("path/to/file.txt"));
+      const expected = Type.list(
+        Array.from("file.txt", (char) => Type.integer(char.codePointAt(0))),
```
Comment (bartblast) [inline test/javascript/erlang/filename_test.mjs]: Could you create a test helper `iolist(string)` in `helpers.mjs` and apply it in other `basename/1` tests?

File: test/javascript/erlang/filename_test.mjs [hunk 13]
```diff
@@ -0,0 +1,225 @@
+"use strict";
```
Comment (bartblast) [inline test/javascript/erlang/filename_test.mjs:1]: The comments from Elixir consistency tests apply here too.

File: test/elixir/hologram/ex_js_consistency/erlang/filename_test.exs (hunk 7 above)
Comment (danirukun) [reply test/elixir/hologram/ex_js_consistency/erlang/filename_test.exs]: You mean for bitstring and atoms? Do we only drop the unit tests that test the deeper conversion mechanics from the Erlang module, e.g. Unicode conversions etc.?

Comment (bartblast) [reply test/elixir/hologram/ex_js_consistency/erlang/filename_test.exs]: We want to test only the logic of `:filename.basename/1`, not deeply test the underlying conversion functions (those should have their own tests).

Our testing strategy should be:

1. **Exercise each code path** in the basename function - we need code coverage of all the logical branches (empty path, single segment, multiple segments, trailing slashes, etc.)

2. **Test each allowed input type** - we need to exercise the code branches for each type (atom, bitstring, etc.). Additionally, since the code uses `maybeSetTextFromBytes()`, we need to test both **text-based and bytes-based input for each type** (I've just noticed this requirement). This ensures both internal representations are handled correctly.

3. **Test error handling** with invalid input types

So the issue with having both:
- "list of code points for atom input"
- "list of code points for atom input with single segment"

...is that they're testing the **same code path** (non-empty atom input) twice. We should test each distinct code path once per type-representation combination, but we don't need to test every behavioral scenario (empty path, single segment, multiple segments, etc.) for every type.

**In practice this means:**
- Core behavioral tests: empty path, single segment, multi-segment, trailing slash, etc. (exercise all code paths)
- For each allowed input type: test both text-based AND bytes-based representations if there may be conversion with `maybeSetTextFromBytes()`
- Invalid input error handling

This gives us full code coverage without combinatorial explosion. Ideally, we'd test all combinations of (behavioral scenarios × input types × representations), but in this function that would mean combinatorial explosion, making the tests unmaintainable.

File: assets/js/erlang/filename.mjs [hunk 14]
```diff
@@ -0,0 +1,56 @@
...
+      }
+
+      const binary = Erlang["iolist_to_binary/1"](filename);
+      Bitstring.maybeSetTextFromBytes(binary);
+      filepathText = binary.text;
+      returnAsCodepoints = true;
+    } else if (Type.isAtom(filename)) {
+      const binary = Erlang["atom_to_binary/1"](filename);
+      Bitstring.maybeSetTextFromBytes(binary);
+      filepathText = binary.text;
```
Comment (bartblast) [inline assets/js/erlang/filename.mjs]: Here, this can be simplified to:
`filepathText = filename.text`

File: test/javascript/support/helpers.mjs [hunk 15]
```diff
@@ -334,3 +334,9 @@ export function registerWebApis() {
 export function waitForEventLoop() {
   return new Promise((resolve) => setTimeout(resolve, 0));
 }
+
+export function iolist(string) {
```
Comment (bartblast) [inline test/javascript/support/helpers.mjs]: Could you order the functions alphabetically?

File: assets/js/erlang/filename.mjs (hunk 14 above)
Comment (danirukun) [reply assets/js/erlang/filename.mjs]: Hmm are you sure? If I do that it ends up being undefined

Comment (bartblast) [reply assets/js/erlang/filename.mjs]: Sorry, I meant `filename.value`.

Comment (danirukun) [reply assets/js/erlang/filename.mjs]: ah thought so, gotcha

**Comment by danirukun** [comment]
If you want before merge I can clean up the commits and squash them a bit.

**Comment by bartblast** [comment]
No need, I squash them during the merge anyway. Thanks for your help @danirukun! :)
